### DISCUSSION TITLE
If render_template_as_native_obj=True SimpleHttpOperator converts header values to integer automatically

### DISCUSSION THREAD

**Discussion by subhosuper** [apache/airflow#26336] [General]
### Apache Airflow version

Other Airflow 2 version

### What happened

If render_template_as_native_obj=True 
SimpleHttpOperator converts header values to integer automatically.
 
`headers = {"Content-Type": "application/json", "accept": "application/json", "Accept-Version": "V2", "OTP": "123345"}` is interpretted as `headers = {"Content-Type": "application/json", "accept": "application/json", "Accept-Version": "V2", "OTP": 123345}`

OTP value has to be of string datatype but characters are numbers.

Airflow automatically converts this OTP value to integer whereas the endpoint wants OTP value as string.
This is the error from the endpoint, this is to explain the error I am getting from airflow 
`requests.exceptions.InvalidHeader: Value for header {OTP: 123345} must be of type str or bytes, not <class 'int'>`

If render_template_as_native_obj=False, then SimpleHttpOperator works just fine but I want to use render_template_as_native_obj=True  for other requirements.

### What you think should happen instead

Expected: 
Please suggest a way to use render_template_as_native_obj=True and be able to send numbers as string in headers of SimpleHttpOperators

### How to reproduce

```
with DAG('mydag', schedule_interval='@daily', default_args=default_args, catchup=False, render_template_as_native_obj=True) as dag:
      task = SimpleHttpOperator(
              task_id = 'taskname',
              method = "POST",
              http_conn_id = "host",
              endpoint = "/endpoint",
              headers = {"Content-Type": "application/json", "accept": "application/json", "Accept-Version": "V2", "OTP": "123345"}',
              data = json.dumps({"csr": "{{ ti.xcom_pull(task_ids='task1', key='return_value').replace(' ', '')}}" }),
              response_check = lambda response: _handle_response_check(response),
              response_filter = lambda response: {"id": response.json()["id"], "token": response.json()["token"], "secret": response.json()["secret"]}
          )
```
      
This is how the task looks like.

### Operating System

Ubuntu 20.04

### Versions of Apache Airflow Providers

apache/airflow:2.3.0 as base image and added few python libraries and stored in a private docker repository.
Using this image with helm.

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

Deploying with helm in Kubernetes cluster k8s

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by subhosuper** [comment]
@kaxil @wolfier

**Comment by subhosuper** [comment]
@potiuk

**Comment by potiuk** [comment]
I do not think this is an issue - this is how native rendering works. There is not much we can do about it. 

But I suggest you to try some workarounds. Likely proper escaping of the first digit with unicode/hex character - "\0x3123345" (not sure if this particular escaping will work so maybe you will have to adapt it) - but I think the general approach should work - @subhosuper  can you please experiment with it and see if it works ? If so, then maybe you can contribute a Doc on how to treat similar issues as a contribution to our documentaiton.

Looking forward to results of your experiment. And converting it into discussion.

**Comment by david-royo-db** [comment]
Hi,

Has anyone found a solution for this?

Our DAG relies on render_template_as_native_obj=True for some of the attributes with complex types, but in one particular place we have a parameterized list of spark properties that need to be passed to the dataproc operator to run a spark job.

As soon as the DAG starts the properties are rendered as a dict of strings just fine. The "rendered template" tab shows everything as strings, this is pending some rendering for some task references like '{{ ti.whatever... }}'.

When the actual task starts running all the spark properties which can be cast as a number are just rendered as native numbers, and the dataproc operator just fails because it expects pairs of strings

This is driving us absolutely crazy. We tried passing numbers with a blank space, because spark seems to swallow that just fine, but those strings are also rendered as ints! We tried the hints about unicode escaping but we haven't been able to make it work in any way

Any idea?

**Reply by david-royo-db** [reply]
Hi,

In case someone faces this situation we fixed it by brute force by making all the spark properties string after the DAG has been rendered

![image](https://github.com/apache/airflow/assets/117661491/e2275744-bcdc-4a46-81d5-f6a168d83734)

A clear victory for our team, a great loss for humanity

**Reply by potiuk** [reply]
I think you are absolutely welcome to improve things. Render_template_as_native_obj is a global per-dag settings that uses https://jinja.palletsprojects.com/en/3.0.x/nativetypes/#jinja2.nativetypes.NativeEnvironment which - indeed - "guesses" the type of the values from their string representation. 

But yeah, I see that you could make a PR for example adding a feature to specify for example which task/parameters should be rendered natively - that would a nice addition. Or have a way to specify exceptions when render_template_as_native_obj is set to True.

THAT would be a truly giant leap for humanity.

PRs for that are most welcome.

**Reply by david-royo-db** [reply]
That sounds like a great idea. Apologies if my message sounded sarcastic or pushy in any way but it was more humorous than anything else. You people are doing a great job :-)

I truly am contemplating the idea of doing a PR for the dataproc operators for the google cloud provider as that would be a quick win. Why try to create a protobuf and fail with weird errors with the spark properties which must be string, when you can cast them back to string? At least there's no side effects I guess? (That needs a though though)

But I'm afraid my python knowledge is **far** below what it would be necessary to contribute with a robust generic solution. But who knows, maybe eventually I get there :-)

**Reply by potiuk** [reply]
No worries. Some day :)
